### PR TITLE
[WIP] Mutter 3.36.9 rebase

### PR DESCRIPTION
**Pull request by clefebvre**
needed:
- https://github.com/linuxmint/cinnamon/pull/10372
- https://github.com/linuxmint/cinnamon-settings-daemon/pull/349
- https://github.com/linuxmint/cinnamon-desktop/pull/204
- https://github.com/linuxmint/cinnamon-control-center/pull/280
- (not mandatory) https://github.com/linuxmint/cjs/pull/101
- https://github.com/linuxmint/mint-themes/pull/380

Some transition notes (of interest to xlet devs mainly):
https://docs.google.com/spreadsheets/d/1AHbuhi5bnlN_O0JP0O-rmgbN50IBHqZm74w6A1TL8Nk/edit?usp=sharing

This is just an experiment.

**Loss of functionality**

The following features would be lost and not brought back:

- Window snapping (most people don't even know it's there or what the difference with tiling is)
- Configurable window animations (no real need, if we need to bring it back, we can simplify it greatly)
- Configurable vsync method (extremely niche)
- Tile detection for prefixing window titles in alt-tab and window list (not important)

The following features would need to be brought back:

- Corner tiling (mutter only has 1/2 tiling, but we can add 1/4 tiling in JS, see https://github.com/emasab/shelltile) - *partly restored, issues will resizing against neighbor windows in certain layouts.*
- Window placement mode - muffin now only has "center" or "not center" -- there's no 'manual' mode. Do we want back?
- Evaluate new keybinding actions - application-switch and -backwards, cycle bindings, 'always on top' - see keybindings.c
- 'resize threshold' setting - a value that affects how much you need to drag a max/tiled window to break free. Overkill?
- ignore hide-titlebar-when-maximized feature
- workspace switching during drag with arrow keys
- sync-to-vblank switch
- ibus support
- Documentation for clutter/muffin/cogl - *see google doc link above for info*.

**Regressions**

- ~Existing wm and desktop settings~ - mostly restored.
- ~Titlebar menu (removed in GNOME 3.13.2 https://github.com/GNOME/mutter/commit/8640982e682fe00de21b2ea7301969a7ea74deee)~ - using clutter menu for now

**Bugs**

- Display settings do not apply

**Pros**

- Marginally snappier
- Cleaner codebase
- Partial wayland support? (starts in 3.10)
- GTK borders for all windows (introduced in 3.16)
- ~More than 3 keyboard layouts at once.~ - postponed for now

**Require adaptation**

- For button events, must use event.get_button() instead of checking event state 

**Filenames differences with 5.0**

gir1.2-meta-muffin-0.0

	./usr/lib/x86_64-linux-gnu/muffin/Meta-0.typelib
	/usr/lib/x86_64-linux-gnu/muffin/Meta-Muffin.0.typelib

libmuffin0

	clutter

		./usr/include/muffin/clutter/clutter/clutter-paint-context.h
		./usr/include/muffin/clutter/clutter/clutter-pick-context.h
		./usr/include/muffin/clutter/clutter/clutter-seat.h
		./usr/include/muffin/clutter/clutter/clutter-stage-view.h

		/usr/include/muffin/clutter-0/cally/cally-texture.h
		/usr/include/muffin/clutter-0/clutter/clutter-device-manager.h
		/usr/include/muffin/clutter-0/clutter/clutter-version.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-animatable.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-animator.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-backend.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-behaviour-depth.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-behaviour-ellipse.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-behaviour.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-behaviour-opacity.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-behaviour-path.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-behaviour-rotate.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-behaviour-scale.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-bin-layout.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-cairo-texture.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-frame-source.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-input-device.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-keysyms.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-list-model.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-main.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-media.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-model.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-score.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-shader.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-stage-manager.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-table-layout.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-texture.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-timeout-pool.h
		/usr/include/muffin/clutter-0/clutter/deprecated/clutter-util.h
		/usr/include/muffin/clutter-0/clutter/x11/clutter-x11-texture-pixmap.h

	COGL

		./usr/include/muffin/cogl/cogl/cogl-depth-state.h
		./usr/include/muffin/cogl/cogl/cogl-dma-buf-handle.h
		./usr/include/muffin/cogl/cogl/cogl-pixel-format.h
		./usr/include/muffin/cogl/cogl/cogl-trace.h
		./usr/include/muffin/cogl/cogl/cogl-wayland-server.h

		/usr/include/muffin/cogl/cogl/cogl-auto-texture.h
		/usr/include/muffin/cogl/cogl/cogl-clutter.h
		/usr/include/muffin/cogl/cogl/cogl-deprecated.h
		/usr/include/muffin/cogl/cogl/cogl-depth-state.h
		/usr/include/muffin/cogl/cogl/cogl-error.h
		/usr/include/muffin/cogl/cogl/cogl-euler.h
		/usr/include/muffin/cogl/cogl/cogl-gles2.h
		/usr/include/muffin/cogl/cogl/cogl-gles2-types.h
		/usr/include/muffin/cogl/cogl/cogl-material-compat.h
		/usr/include/muffin/cogl/cogl/cogl-muffin.h
		/usr/include/muffin/cogl/cogl/cogl-pango.h
		/usr/include/muffin/cogl/cogl/cogl-primitives.h
		/usr/include/muffin/cogl/cogl/cogl-quaternion.h
		/usr/include/muffin/cogl/cogl/cogl-shader.h
		/usr/include/muffin/cogl/cogl/cogl-texture-2d-gl.h
		/usr/include/muffin/cogl/cogl/cogl-texture-3d.h
		/usr/include/muffin/cogl/cogl/cogl-texture-rectangle.h
		/usr/include/muffin/cogl/cogl/cogl-type-casts.h
		/usr/include/muffin/cogl/cogl/cogl-vector.h
		/usr/include/muffin/cogl/cogl/cogl-vertex-buffer.h
		/usr/include/muffin/cogl/cogl/deprecated/cogl-clutter-xlib.h
		/usr/include/muffin/cogl/cogl/deprecated/cogl-framebuffer-deprecated.h
		/usr/include/muffin/cogl/cogl/deprecated/cogl-vertex-buffer.h

	Meta

		./usr/include/muffin/meta/barrier.h
		./usr/include/muffin/meta/compositor-mutter.h
		./usr/include/muffin/meta/meta-backend.h
		./usr/include/muffin/meta/meta-background-group.h
		./usr/include/muffin/meta/meta-background.h
		./usr/include/muffin/meta/meta-background-image.h
		./usr/include/muffin/meta/meta-close-dialog.h
		./usr/include/muffin/meta/meta-cursor-tracker.h
		./usr/include/muffin/meta/meta-dnd.h
		./usr/include/muffin/meta/meta-enum-types.h
		./usr/include/muffin/meta/meta-idle-monitor.h
		./usr/include/muffin/meta/meta-inhibit-shortcuts-dialog.h
		./usr/include/muffin/meta/meta-launch-context.h
		./usr/include/muffin/meta/meta-monitor-manager.h
		./usr/include/muffin/meta/meta-remote-access-controller.h
		./usr/include/muffin/meta/meta-selection.h
		./usr/include/muffin/meta/meta-selection-source.h
		./usr/include/muffin/meta/meta-selection-source-memory.h
		./usr/include/muffin/meta/meta-settings.h
		./usr/include/muffin/meta/meta-sound-player.h
		./usr/include/muffin/meta/meta-stage.h
		./usr/include/muffin/meta/meta-startup-notification.h
		./usr/include/muffin/meta/meta-version.h
		./usr/include/muffin/meta/meta-window-group.h
		./usr/include/muffin/meta/meta-window-shape.h
		./usr/include/muffin/meta/meta-workspace-manager.h
		./usr/include/muffin/meta/meta-x11-display.h
		./usr/include/muffin/meta/meta-x11-errors.h

		/usr/include/muffin/meta/atomnames.h
		/usr/include/muffin/meta/compositor-muffin.h
		/usr/include/muffin/meta/errors.h
		/usr/include/muffin/meta/gradient.h
		/usr/include/muffin/meta/preview-widget.h
		/usr/include/muffin/meta/screen.h

	SO

		./usr/lib/x86_64-linux-gnu/muffin/libmuffin-clutter-0.so.0
		./usr/lib/x86_64-linux-gnu/muffin/libmuffin-clutter-0.so.0.0.0
		./usr/lib/x86_64-linux-gnu/muffin/libmuffin-cogl-0.so.0
		./usr/lib/x86_64-linux-gnu/muffin/libmuffin-cogl-0.so.0.0.0
		./usr/lib/x86_64-linux-gnu/muffin/libmuffin-cogl-pango-0.so.0
		./usr/lib/x86_64-linux-gnu/muffin/libmuffin-cogl-pango-0.so.0.0.0
		./usr/lib/x86_64-linux-gnu/muffin/libmuffin-cogl-path-0.so.0
		./usr/lib/x86_64-linux-gnu/muffin/libmuffin-cogl-path-0.so.0.0.0
		./usr/lib/x86_64-linux-gnu/muffin/Meta-0.gir
		./usr/lib/x86_64-linux-gnu/pkgconfig/libmuffin-0.pc

		/usr/lib/x86_64-linux-gnu/libmuffin-clutter-0.so
		/usr/lib/x86_64-linux-gnu/libmuffin-cogl-0.so
		/usr/lib/x86_64-linux-gnu/libmuffin-cogl-pango-0.so
		/usr/lib/x86_64-linux-gnu/libmuffin-cogl-path-0.so
		/usr/lib/x86_64-linux-gnu/muffin/Meta-Muffin.0.gir
		/usr/lib/x86_64-linux-gnu/pkgconfig/libmuffin.pc
		/usr/lib/x86_64-linux-gnu/pkgconfig/muffin-plugins.pc

muffin-common

       OK

muffin

	OK

muffin-dbg

	OK

### CODE REVIEW

**Comment by Fantu** [comment]
@smcv @lestcape have you any advices that can help in this big task please?

**Comment by leigh123linux** [comment]
> Hi Fantu,
> 
> We don't really use or expose muffin (the binary) to users. It's only used by us to troubleshoot issues and we never had a man page for it before.

There is a man page for muffin, see https://koji.fedoraproject.org/koji/fileinfo?rpmID=27046156&filename=/usr/share/man/man1/muffin.1.gz

**Comment by Fantu** [comment]
@clefebvre thanks, take a look to https://github.com/linuxmint/cinnamon/pull/10377 to make cinnamon part of this task building also in a clean environments

**Comment by Fantu** [comment]
I added to my commit also the missed breaks/replace in -dev split to make working also upgrade (sorry for mistake), I din't add version (<< 5.2.0) so will be working also from existing 5.2 testing build (can be added after PR merging)
https://github.com/Fantu/muffin/commits/fix-5.2-test

**Comment by Fantu** [comment]
@clefebvre thanks for resplit the -dev package, is missed the cinnamon part, here updated build-deps/deps and meson dep: https://github.com/linuxmint/cinnamon/pull/10385

for people that want do a fast test I prepared a ppa, I tested on mint 20.2, added linux mint nightly build ppa and my ppa
mint ppa have highter pinning to another changes is needed for install/upgrade cinnamon/muffin from my ppa:

```
sudo add-apt-repository ppa:linuxmint-daily-build-team/daily-builds
sudo add-apt-repository ppa:fantonifabio/muffin-rebase
sudo apt-get update
sudo nano /etc/apt/preferences.d/fantu.pref

Package: *
Pin: release o=LP-PPA-fantonifabio-muffin-rebase
Pin-Priority: 700

sudo apt update
```

**Comment by clefebvre** [comment]
@Fantu thanks, I wouldn't worry about csettings just yet. There are so many bugs with the core desktop still, one thing at a time :)

**Comment by Fantu** [comment]
I updated my ppa with latest muffin/cinnamon fixes and did another fast test (only in a vm for now)
In .xsession-errors I see many of:
```
(cinnamon:1198): Clutter-CRITICAL **: 21:55:55.404: clutter_actor_get_transformed_size: assertion 'CLUTTER_IS_ACTOR (self)' failed

(cinnamon:1198): Cjs-CRITICAL **: 21:55:55.404: Object Clutter.Actor (0x55c9d6879410), has been already deallocated — impossible to access it. This might be caused by the object having been destroyed from C code using something such as destroy(), dispose(), or remove() vfuncs.
== Stack trace for context 0x55c9d617f6a0 ==
#0   7ffeac7b8d40 b   /usr/share/cinnamon/js/ui/layout.js:725 (3f2a6b65cd80 @ 1098)
#1   7ffeac7b8e00 b   self-hosted:850 (3d32955c3bf0 @ 398)

(cinnamon:1198): Clutter-CRITICAL **: 21:55:55.405: clutter_actor_get_paint_visibility: assertion 'CLUTTER_IS_ACTOR (actor)' failed
```
I don't have time now to do a check about it if related to muffin rebase and if I can do a fix, I posted if can be useful, if you need a specific test for help or something else try to tell and I'll do if I'll have enough time, for now I'll try at least to update ppa for anyone want test it in easy/fast way
@clefebvre @mtwebster @JosephMcc thanks for your works

**Comment by JosephMcc** [comment]
@Fantu I think some of those are coming from an issue with the backgrounds. When I first start I get a black background and those errors. But if I restart cinnamon, I get a proper background and those errors seem to go away.

**Comment by Fantu** [comment]
@letscape uma have already daily build packages by default? I saw that kelebek333 ppa don't depend on dailybuild and don't have full packages updated from git in it, on my ppa I keep daily build as dep for rebuild only cinnamon and muffin for this task (like when this will be merged and also because cinnamon also include other commits for next version) and I wrote instructions for use it on a post above (now I'll also include in ppa description to avoid mistake). and about my additional commits I did only few small things to make cinnamon build and very small packaging improvements awaiting that will be merged, should be nothing that change the tests, or I'm missing something?

**Comment by lestcape** [comment]
This is my proposition to fix the vector-box _keepMaskActive jumping issue:

The reason of why _keepMaskActive is jumping between true and false is because the current code is enabled the VectorMask without to really check if is needed we need to enable is, but to check that we first need to have a the information of the  vector_mask (this.vector_mask_info) to then check it, so the _categoryMotionEvent function should be convert to something like this:
```js
_categoryMotionEvent(actor, event) {
    // Always keep the mask engaged - motion-events on the category buttons
    // trigger this.

    if (this.vector_update_loop == 0) {
        if (!this.vector_mask_info) {
            this.vector_mask_info = this._getNewVectorInfo(actor);
        }
        if (this._keepMaskActive()) {
            this._enableVectorMask(actor);
        }
    }

    return Clutter.EVENT_PROPAGATE;
}
```
Unfortunately, this is not enough, because there are another problem. The problem is that we are using the broken function calc_angle to finding our angle direction. The broken part of the function is here:

```js
if (x == 0 || y == 0) {
    return 0;
}
```
because in our cartesian system references when y == 0 the angle is not always 0 because depend of the sign of x it can be 0 or 180 degrees. The same occurs when x == 0 the angle is not 0 is instead 90 or -90 degrees depending of the y value, if is positive or negative. So to fix that we need to replace the calc_angle function with this one:
```js
function calc_angle(x, y) {
    let r = Math.atan2(y, x) * (180 / Math.PI);
    return r;
}
```
Please note that Math.atan2 is giving us the right value automatically, so the right thing to do here is not do anything more.

Here, the problem of the menu was  gone, but i think that we also can add some improvements. For example:
The logs are done that way:
```js
log(`${this.vector_mask_info.top_angle.toFixed()} <---${angle.toFixed()}---> ${this.vector_mask_info.bottom_angle.toFixed()} - Continue? ${ret}`);
```
and this not get us an easy way to know if the angle is negative or positive because the separator <--- have the same character of the negative sign.

Another improve is just to add a gap to our angle to ensure that the corner cases do not activate the vectorbox. I think that 10 degrees is enough. So, we can then check if our angle is in range in this way instead:
```js
ret = angle <= this.vector_mask_info.top_angle + 10 &&
      angle >= this.vector_mask_info.bottom_angle - 10;
```

**Comment by Heston-in-Mtl** [comment]
Thank you @clefebvre for your continued maintenance of Mint/Cinnamon!  I'm a multiple time $ contributor to the project.

Not sure if this is the appropriate venue for this, but I just wanted to cast my vote on 'resize threshold' setting.  I do use and appreciate this setting.  I find it especially useful with low-precision laptop touchpads.
Cheers!

**Comment by Fantu** [comment]
@mtwebster build on focal fails for missed newer pipewire required by remote desktop so I think is good keep this for now
`-Dremote_desktop=false`

File: debian/rules [hunk 1]
```diff
@@ -11,7 +11,7 @@ override_dh_autoreconf:
 
 CONFFLAGS = \
 	-Dprofiler=false \
-	-Ddefault_driver=gl3
+	-Ddefault_driver=gl
```
Comment (mtwebster) [inline debian/rules:14]: @Fantu sorry I was curious about building with gl3 and forgot to change it back before committing.

**Comment by mtwebster** [comment]
Ok disabled that again.

I'm going to work on rebasing muffin and cinnamon against master. I shouldn't have waited so long..

**Comment by Fantu** [comment]
@mtwebster thanks for your big works, if confirmed muffin rebase for cinnamon 5.4 I think can be better merge to master soon to avoid waste time on future rebase your and of all contributors that will still make PR based on master. it will make also possible more testers since now and this I think will make possible improve quality of next version, this is also a big rebase after 10 years, so more testers and time to tests will be available is better

**Comment by Fantu** [comment]
@mtwebster thanks for your works
I tested latest commits (https://launchpad.net/~fantonifabio/+archive/ubuntu/muffin-rebase), done 2 PR with some debian improvements: https://github.com/linuxmint/cinnamon-control-center/pull/281 https://github.com/linuxmint/cinnamon-desktop/pull/206 and I spotted something: [screenshot](https://ibb.co/51MPhZy)
opening display setting show an error in terminal, I'm unable to place the display settings window in the middle/right of the screen (holding down the left button and bringing it to the right side, with the terminal window on middle/left instead worked correctly), titlebar of terminal window do an issue when I select other window, all things are visible in screenshot above.
Hoping that these small informations from my test can be useful, sorry if I don't have enough time to help more and for my bad english.

**Comment by ItzSwirlz** [comment]
Nice!

@Fantu you said you were going to adjust the hardening in Debian?

**Comment by Fantu** [comment]
@ItzSwirlz as you can see debian/rules of muffin rebase there is already all hardening enabled (same of mutter), now that code are improved with the rebase and on tests I did for now I not found issue related so will be applied also to muffin 5.4 debian package

**Comment by lestcape** [comment]
Also you probably want to fork mutter again to a point where you can add this: https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/1441 Have 60 FPS instead of 30 FPS in x11 seem to be as very good improve.

**Pull request merged**